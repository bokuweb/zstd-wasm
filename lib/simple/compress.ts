import { waitInitialized, Module } from '../init';
import { isError, getErrorName } from '../errors';

const compressBound = (size: number): number => {
  const bound = Module.cwrap('ZSTD_compressBound', 'number', ['number']);
  return bound(size);
};

export const compress = async (buf: ArrayBuffer, level: number) => {
  await waitInitialized();
  const bound = compressBound(buf.byteLength);
  const malloc = Module.cwrap('create_buffer', 'number', ['number']);
  const compressed = malloc(bound);
  const free = Module.cwrap('destroy_buffer', 'number');
  try {
    /*
      @See https://zstd.docsforge.com/dev/api/ZSTD_compress/
      size_t ZSTD_compress( void* dst, size_t dstCapacity, const void* src, size_t srcSize, int compressionLevel);
      Compresses `src` content as a single zstd compressed frame into already allocated `dst`.
      Hint : compression runs faster if `dstCapacity` >=  `ZSTD_compressBound(srcSize)`.
      @return : compressed size written into `dst` (<= `dstCapacity),
                or an error code if it fails (which can be tested using ZSTD_isError()).
    */
    const _compress = Module.cwrap('ZSTD_compress', 'number', ['number', 'number', 'array', 'number', 'number']);
    const sizeOrError = _compress(compressed, bound, buf, buf.byteLength, level);
    if (isError(sizeOrError)) {
      throw new Error(getErrorName(sizeOrError));
    }
    // // Copy buffer
    // // Uint8Array.prototype.slice() return copied buffer.
    const data = new Uint8Array(Module.HEAPU8.buffer, compressed, sizeOrError).slice();
    free(compressed, bound);
    return data;
  } catch (e) {
    free(compressed, bound);
    throw e;
  }
};
