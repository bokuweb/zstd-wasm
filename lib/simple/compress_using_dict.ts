import { Module } from '../module';
import { isError } from '../errors';

const compressBound = (size: number): number => {
  const bound = Module['_ZSTD_compressBound'];
  return bound(size);
};

export const createCCtx = (): number => {
  return Module['_ZSTD_createCCtx']();
};

export const compressUsingDict = (cctx: number, buf: ArrayBuffer, dict: ArrayBuffer, level?: number) => {
  const bound = compressBound(buf.byteLength);
  const malloc = Module['_malloc'];
  const compressed = malloc(bound);
  const src = malloc(buf.byteLength);
  Module.HEAP8.set(buf, src);
  // Setup dict
  const pdict = malloc(dict.byteLength);
  Module.HEAP8.set(dict, pdict);
  const free = Module['_free'];
  try {
    /*
      @See https://zstd.docsforge.com/dev/api/ZSTD_compress_usingDict/
      size_t ZSTD_compress_usingDict(ZSTD_CCtx* cctx,
                         void* dst, size_t dstCapacity,
                         const void* src, size_t srcSize,
                         const void* dict, size_t dictSize,
                         int compressionLevel)
    */
    const _compress = Module['_ZSTD_compress_usingDict'];
    const sizeOrError = _compress(cctx, compressed, bound, src, buf.byteLength, pdict, dict.byteLength, level ?? 3);
    if (isError(sizeOrError)) {
      throw new Error(`Failed to compress with code ${sizeOrError}`);
    }
    // // Copy buffer
    // // Uint8Array.prototype.slice() return copied buffer.
    const data = new Uint8Array(Module.HEAPU8.buffer, compressed, sizeOrError).slice();
    free(compressed, bound);
    free(src, buf.byteLength);
    free(pdict, dict.byteLength);
    return data;
  } catch (e) {
    free(compressed, bound);
    free(src, buf.byteLength);
    free(pdict, dict.byteLength);
    throw e;
  }
};
