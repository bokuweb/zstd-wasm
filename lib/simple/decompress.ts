import { Module } from '../module';
import { isError } from '../errors';

const getFrameContentSize = (src: number, size: number): number => {
  const getSize = Module['_ZSTD_getFrameContentSize'];
  return getSize(src, size);
};

export type DecompressOption = {
  defaultHeapSize?: number;
};

export const decompress = (
  buf: Uint8Array,
  opts: DecompressOption = { defaultHeapSize: 1024 * 1024 }, // Use 1MB on default if it is failed to get content size.
): Uint8Array => {
  const malloc = Module['_malloc'];
  const src = malloc(buf.byteLength);
  Module.HEAP8.set(buf, src);
  const contentSize = getFrameContentSize(src, buf.byteLength);
  const size = contentSize === -1 ? opts.defaultHeapSize : contentSize;
  const free = Module['_free'];
  const heap = malloc(size);
  try {
    /*
      @See https://zstd.docsforge.com/dev/api/ZSTD_decompress/
      compressedSize : must be the exact size of some number of compressed and/or skippable frames.
      dstCapacity is an upper bound of originalSize to regenerate.
      If user cannot imply a maximum upper bound, it's better to use streaming mode to decompress data.
      @return: the number of bytes decompressed into dst (<= dstCapacity), or an errorCode if it fails (which can be tested using ZSTD_isError()).
    */
    const _decompress = Module['_ZSTD_decompress'];
    const sizeOrError = _decompress(heap, size, src, buf.byteLength);
    if (isError(sizeOrError)) {
      throw new Error(`Failed to compress with code ${sizeOrError}`);
    }
    // Copy buffer
    // Uint8Array.prototype.slice() return copied buffer.
    const data = new Uint8Array(Module.HEAPU8.buffer, heap, sizeOrError).slice();
    free(heap, size);
    free(src, buf.byteLength);
    return data;
  } catch (e) {
    free(heap, size);
    free(src, buf.byteLength);
    throw e;
  }
};
