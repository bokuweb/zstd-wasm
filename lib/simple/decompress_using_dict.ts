import { Module } from '../module';
import { isError } from '../errors';
import { DecompressOption } from './decompress';

const getFrameContentSize = (src: number, size: number): number => {
  const getSize = Module['_ZSTD_getFrameContentSize'];
  return getSize(src, size);
};

export const createDCtx = (): number => {
  return Module['_ZSTD_createDCtx']();
};

export const freeDCtx = (dctx: number) => {
  return Module['_ZSTD_freeDCtx'](dctx);
};

export const decompressUsingDict = (
  dctx: number,
  buf: ArrayBuffer,
  dict: ArrayBuffer,
  opts: DecompressOption = { defaultHeapSize: 1024 * 1024 }, // Use 1MB on default if it is failed to get content size.
): ArrayBuffer => {
  const malloc = Module['_malloc'];
  const src = malloc(buf.byteLength);
  Module.HEAP8.set(buf, src);
  const pdict = malloc(dict.byteLength);
  Module.HEAP8.set(dict, pdict);
  const contentSize = getFrameContentSize(src, buf.byteLength);
  const size = contentSize === -1 ? opts.defaultHeapSize : contentSize;
  const free = Module['_free'];
  const heap = malloc(size);
  try {
    const _decompress = Module['_ZSTD_decompress_usingDict'];
    const sizeOrError = _decompress(dctx, heap, size, src, buf.byteLength, pdict, dict.byteLength);
    if (isError(sizeOrError)) {
      throw new Error(`Failed to compress with code ${sizeOrError}`);
    }
    // Copy buffer
    // Uint8Array.prototype.slice() return copied buffer.
    const data = new Uint8Array(Module.HEAPU8.buffer, heap, sizeOrError).slice();
    free(heap, size);
    free(src, buf.byteLength);
    free(pdict, dict.byteLength);
    return data;
  } catch (e) {
    free(heap, size);
    free(src, buf.byteLength);
    free(pdict, dict.byteLength);
    throw e;
  }
};
