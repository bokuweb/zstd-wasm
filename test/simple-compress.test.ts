import { decompress } from '../lib/simple/decompress';
import { compress } from '../lib/simple/compress';

test('hello', async () => {
  const compressed = await compress(Buffer.from('Hello'), 10);
  expect(compressed).toMatchSnapshot();
  const decompressed = await decompress(compressed);
  expect(Buffer.from(decompressed).toString()).toBe('Hello');
});
