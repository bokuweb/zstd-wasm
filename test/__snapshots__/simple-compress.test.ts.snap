// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hello 1`] = `
Uint8Array [
  40,
  181,
  47,
  253,
  32,
  5,
  41,
  0,
  0,
  72,
  101,
  108,
  108,
  111,
]
`;
